plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,504 @@
+                                                       )
+              let (mAnn, mRecurs, newBindings) = foldr1 (<>) $ fmap (\(oAnn,oRec,oBinding,_) -> (oAnn
+                                                                                     -- if the SCC is a single node then use its original 'Recursivity';
+                                                                                     -- otherwise, the SCC is  a group of nodes and we *have to* treat all of them in a 'letrec',
+                                                                                     -- since we don't have any information on how to linearize those.
+                                                                                   , if isSingleton scc then oRec else Rec
+                                                                                     -- lift the binding into a semigroup for accumulation
+                                                                                   , pure oBinding)
+                                                                    ) res
+              pure $ case accTerm of
+                       -- MERGE current let-group with previous let-group iff both groups' recursivity is NonRec
+                       Let accAnn NonRec accBs accIn | mRecurs == NonRec -> Let (mAnn <> accAnn) NonRec (newBindings <> accBs) accIn
+                       -- never merge if the previous let-group is a Rec or the current let-group is Rec,
+                       -- but instead create a nested current let-group under the previous let-group (above)
+                       _ -> Let mAnn mRecurs newBindings accTerm
+
+
+-- | The main transformation function (Term -> Term) that returns a term with all lets floated as outwards as possible (full laziness).
+-- Is comprised of two AST "passes":
+-- 1stpass: to collect the ranks (positions) of all lets
+-- 2ndpass:  to remove all its lets and place them back (float them) in their ranks (positions).
+-- See Note [Float algorithm]
+--
+-- NB: This transformation requires that the PLC.rename compiler-pass has prior been run.
+floatTerm :: forall name tyname uni a
+           . (PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique, Monoid a)
+          => Term tyname name uni a -> Term tyname name uni a
+floatTerm pir = p2Term pir
+                -- give the floatdata of the 1st pass to the start of the 2nd pass
+              $ p1Term pir
+
+
+
+-- Helpers
+----------
+
+-- | A getter that returns a single 'Unique' for a particular binding.
+-- We need this because let-datatypes introduce multiple identifiers, but in our 'RhsTable', we use a single Unique as the key. See also: 'bindingIds'
+principal :: (PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique)
+            => Getting r (Binding tyname name uni a) PLC.Unique
+principal = to $ \case TermBind _ _ (VarDecl _ n _) _ -> n ^. PLC.theUnique
```

Comment:
Sneaky suggestion to save some code: `min $ bindingIds b` or something. This won't always pick the constructor, but since it's arbitrary, that doesn't matter.

Reply:
I personally like the current less-arbitrary version more, but whatever.